### ISSUE TITLE
[Head] Add π0 flow-matching action expert

### ISSUE THREAD

**Issue by elprofesoriqo** [elprofesoriqo/FlowEdge#11]
## Goal

Add the π0 flow-matching action expert to FlowEdge.

The initial implementation is **head-only**. PaliGemma image and language encoding stays outside FlowEdge. The runtime receives a precomputed prefix representation and executes the action expert plus flow-matching integration.

Reference checkpoint:

* `lerobot/pi0_base`

## Scope

Support:

* π0 Gemma action expert
* precomputed VLM prefix input
* robot state conditioning
* noisy action input
* timestep conditioning
* action chunk generation
* Euler flow integration
* FP32
* BF16 where supported by existing kernels
* action normalization and un-normalization
* C API
* Python API

The initial LeRobot configuration uses a `gemma_300m` action expert, action horizon `50`, maximum state dimension `32`, maximum action dimension `32`, and `10` inference steps by default.

## Out of scope

Not required for the first implementation:

* PaliGemma vision encoder
* image preprocessing
* language tokenization
* full VLM inference
* π0-FAST
* π0.5
* training or fine-tuning
* dynamic action horizon
* arbitrary Gemma variants
* accelerator-specific kernels

## Architecture

```mermaid
flowchart TD
    A[Precomputed VLM prefix] --> F[Prefix KV state]

    B[Robot state] --> C[State projection]

    D[Noisy action chunk x_t] --> E[Action projection]
    G[Timestep t] --> H[Timestep embedding]

    C --> I[Action/time conditioning]
    E --> I
    H --> I

    F --> J[π0 Gemma action expert]
    I --> J

    J --> K[Action output projection]
    K --> L[Velocity field v_t]

    L --> M[Euler update]
    M --> N{More flow steps?}

    N -->|Yes| D
    N -->|No| O[Normalized action chunk]

    O --> P[Un-normalize actions]
    P --> Q[Robot actions]
```

## Action expert

Add:

```text
heads/pi0/
├── pi0.h
└── pi0.cc
```

Implement the π0 action expert using the Transformer primitives shared with #10 where possible.

The action expert should support:

* [ ] Gemma-style attention
* [ ] grouped key/value heads
* [ ] RMSNorm
* [ ] GELU-tanh MLP activation
* [ ] fixed-size attention state
* [ ] precomputed prefix KV input
* [ ] action-token processing
* [ ] state and timestep conditioning

The initial `gemma_300m` expert uses:

```text
hidden_size = 1024
layers = 18
attention_heads = 8
kv_heads = 1
head_dim = 256
mlp_dim = 4096
```

These values should come from checkpoint metadata rather than being hardcoded into execution code.

## Prefix interface

FlowEdge should not run PaliGemma in this issue.

Instead, define an explicit input contract for the precomputed prefix.

```mermaid
flowchart LR
    A[External PaliGemma / VLM] --> B[Prefix representation]
    B --> C[FlowEdge π0 head]

    D[Robot state] --> C
    E[Initial action noise] --> C

    C --> F[Action chunk]
```

The interface should define:

* [ ] prefix tensor shape
* [ ] prefix mask
* [ ] position information if required
* [ ] prefix KV-cache layout
* [ ] supported precision
* [ ] maximum prefix length

The prefix representation must be documented well enough that another runtime can generate it without depending on FlowEdge internals.

## State and action projections

Implement the projections used by the action expert:

* [ ] `state_proj`
* [ ] `action_in_proj`
* [ ] `action_out_proj`
* [ ] action/time MLP input projection
* [ ] action/time MLP output projection

The reference π0 implementation projects state and action tensors into the action expert width before the Transformer path and projects the final suffix representation back to action dimension.

## Timestep conditioning

Implement the timestep path used by π0:

```mermaid
flowchart LR
    A[Timestep t] --> B[Sinusoidal embedding]
    B --> C[Concatenate / combine with action state]
    C --> D[MLP in]
    D --> E[Activation]
    E --> F[MLP out]
    F --> G[Action expert conditioning]
```

Requirements:

* [ ] deterministic timestep embedding
* [ ] fixed-size working buffers
* [ ] reference parity test

## Flow integration

Reuse the existing FlowEdge flow solver infrastructure where the π0 update rule matches the current Euler path.

```mermaid
flowchart TD
    A[Initial Gaussian action noise x_0] --> B[π0 velocity field]
    B --> C[Euler step]
    C --> D[x_t]
    D --> E{Final step?}

    E -->|No| B
    E -->|Yes| F[Action chunk]
```

Initial support:

* Euler only
* fixed number of inference steps
* default reference configuration of 10 steps

Requirements:

* [ ] no heap allocation inside the integration loop
* [ ] deterministic output for fixed input and initial noise
* [ ] timestep schedule matches the reference implementation

## Normalization

Carry required action statistics through conversion.

The runtime should:

* [ ] load action normalization metadata
* [ ] generate actions in normalized space
* [ ] un-normalize before returning the final action chunk
* [ ] validate action dimension against checkpoint metadata

## Model conversion

Extend `convert/` with an explicit π0 action-expert conversion path.

```mermaid
flowchart LR
    A[LeRobot π0 checkpoint] --> B[FlowEdge converter]

    B --> C[Skip PaliGemma weights]
    B --> D[Action expert weights]
    B --> E[State/action projections]
    B --> F[Timestep MLP]
    B --> G[Normalization metadata]
    B --> H[π0 configuration]

    D --> I[FlowEdge checkpoint]
    E --> I
    F --> I
    G --> I
    H --> I
```

The converter should:

* [ ] import only the weights needed for head-only inference
* [ ] skip PaliGemma weights explicitly
* [ ] map Gemma expert layers
* [ ] map state/action projections
* [ ] map timestep-conditioning projections
* [ ] preserve action horizon
* [ ] preserve action dimension
* [ ] preserve normalization metadata
* [ ] reject unsupported π0 variants

## Verification

Extend `torch_ref.py` with a head-only π0 reference path.

Use fixed:

* prefix representation
* robot state
* initial action noise
* timestep schedule

Verify:

* [ ] state projection
* [ ] action projection
* [ ] timestep embedding
* [ ] single action-expert layer
* [ ] full velocity prediction
* [ ] individual Euler step
* [ ] complete action trajectory
* [ ] final un-normalized action output

### End-to-end parity

```mermaid
flowchart TD
    A[Fixed prefix + state + noise]

    A --> B[LeRobot reference]
    A --> C[FlowEdge π0]

    B --> D[Reference action chunk]
    C --> E[FlowEdge action chunk]

    D --> F{Within tolerance?}
    E --> F

    F -->|Yes| G[Pass]
    F -->|No| H[Fail]
```

## Public API

Expose:

* [ ] π0 model loading
* [ ] prefix input
* [ ] state input
* [ ] optional initial noise for deterministic testing
* [ ] action horizon reporting
* [ ] action dimension reporting
* [ ] inference step configuration
* [ ] C API
* [ ] Python binding

## Tests

* [ ] Gemma RMSNorm vs reference
* [ ] grouped KV attention vs reference
* [ ] state projection parity
* [ ] action projection parity
* [ ] timestep embedding parity
* [ ] action/time MLP parity
* [ ] action expert block parity
* [ ] velocity output parity
* [ ] Euler step parity
* [ ] fixed-noise deterministic inference
* [ ] normalization / un-normalization
* [ ] converter validation
* [ ] unsupported variant rejection
* [ ] no-allocation sampling loop

## Performance requirements

After initialization:

* no heap allocation during action generation
* prefix state stays fixed
* working buffers are reused
* action and timestep buffers are preallocated

Benchmark:

* one action-expert forward
* one Euler step
* complete action chunk generation

Report:

* mean
* p50
* p95
* p99

Include model configuration, prefix length, action horizon, precision, CPU, compiler, and build type.

## Acceptance criteria

* [ ] `lerobot/pi0_base` action-expert weights convert successfully
* [ ] PaliGemma weights are excluded
* [ ] converted checkpoint loads in FlowEdge
* [ ] precomputed prefix input is documented
* [ ] action-expert velocity output matches the reference within documented tolerance
* [ ] complete fixed-noise trajectory matches the reference within documented tolerance
* [ ] Euler integration performs no heap allocation
* [ ] action normalization metadata is preserved
* [ ] returned actions are correctly un-normalized
* [ ] C API works
* [ ] Python API works
* [ ] tests pass
* [ ] supported scope and limitations are documented

## Dependencies

Depends on reusable Transformer primitives from #10.

π0 also needs Gemma-specific pieces not required by the initial GPT-2 scope in #10:

* RMSNorm
* grouped KV attention
* Gemma MLP behavior

These should either be added as child issues of #10 or tracked as π0-specific prerequisites here.

## Suggested child issues

* [ ] `[Kernel] Add RMSNorm and Gemma GELU-tanh`
* [ ] `[Kernel] Add grouped KV attention support`
* [ ] `[Head] Implement π0 state, action and timestep projections`
* [ ] `[Head] Implement π0 Gemma action expert`
* [ ] `[Convert] Add lerobot/pi0_base head-only converter`
* [ ] `[Verify] Add π0 fixed-prefix parity fixtures`

## References

* `lerobot/pi0_base`
* LeRobot π0 implementation
* #10 causal Transformer backbone

**Comment by elprofesoriqo** [comment]
Deferred until the SmolVLA cached expert is a LeRobot policy type and Diffusion Policy beats PyTorch on the published CPU fixture.
